modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/resources/META-INF/resources/js/reducers/defaultReducer.es.js

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {EVENT_TYPES} from '../actions/types.es';
+import {PagesVisitor} from '../util/visitors.es';
+
+export default (state, action) => {
+	switch (action.type) {
+		case EVENT_TYPES.ALL: {
+			const {defaultLanguageId} = state;
+			const {editingLanguageId, pages} = action.payload;
+
+			if (
+				editingLanguageId &&
+				state.editingLanguageId !== editingLanguageId
+			) {
+				const visitor = new PagesVisitor(pages ?? action.pages);
+
+				return {
+					...action.payload,
+					pages: visitor.mapFields(
+						({localizedValue}) => {
+							let value;
+
+							if (localizedValue) {
+								if (localizedValue[editingLanguageId.newVal]) {
+									value =
+										localizedValue[
+											editingLanguageId.newVal
+										];
+								}
+								else if (localizedValue[defaultLanguageId]) {
+									value = localizedValue[defaultLanguageId];
+								}
+							}
+
+							return {
+								value,
+							};
+						},
+						true,
+						true,
+						true
```

Comment:
I have no idea what these do without looking at another file. Would be nice if we could make things a little more transparent.

Reply:
Yeah, I was confused when I was using it. This comes from the visitors that already existed, if I change it I have to change it in many files that depend a lot on it.